Change 'proxy_protocol' default mode and behavior

AntonAM opened the pull request:
Now by default `proxy_protocol` is working in 'unspecified' mode and in that mode we mark incoming connection with setting source port = 0 and don't allow IP pinned connections (based on port = 0).
Also `on` mode now requires PROXY header.

Implements RFD146 https://github.com/gravitational/teleport-private/pull/991

AntonAM committed:
Change 'proxy_protocol' default mode and behavior

Now by default `proxy_protocol` is unspecified and in that mode
we don't allow IP pinned connections and mark incoming conection with setting
source port = 0.
'on' mode now requires PROXY header.

tigrato commented on lib/config/fileconf.go:
```diff
@@ -989,9 +989,13 @@ func (s *Service) Disabled() bool {
+	// Possible values:
+	// - 'on': one PROXY header is accepted and required per incoming connection.
+	// - 'off': no PROXY headers are allows, otherwise connection is rejected.
+	// If unspecified - one PROXY header is allowed, but not required. Connection is marked with source port set to 0.
```
Please add a note that the proxy protocol not being specified is incorrect and IP pinning won't work.
It's also important to mention that you should configure it based on your L4 configs if there is a L4 at all.

AntonAM replied:
I've made same description here and in multiplexer, adding the details you've mentioned.

tigrato commented on lib/multiplexer/multiplexer.go:
```diff
@@ -53,6 +53,19 @@ var (
+// Possible values:
+// - 'on': one PROXY header is accepted and required per incoming connection.
+// - 'off': no PROXY headers are allows, otherwise connection is rejected.
+// - (unspecified): one PROXY header is allowed, but not required. Connection is marked with source port set to 0.
```
Refer why we added this mode and that it's incorrect and no-one should ever use it.
Either use on if your L4 LB sends proxy lines or off if you don't send proxy lines at all or no L4 load balancer exists

tigrato commented on lib/multiplexer/multiplexer.go:
```diff
@@ -414,75 +433,102 @@ func (m *Mux) detect(conn net.Conn) (*Conn, error) {
+			unsignedPROXYLineReceived = true
+
+			if m.PROXYProtocolMode == PROXYProtocolUnspecified {
+				m.Logger.WithFields(log.Fields{
```
I think we should rate this log a bit otherwise we're going to spam a lot.

AntonAM replied:
I personally am ok with spamming it, that's kinda the point. But I guess we can rate limit them as well. We already have log limiter in the multiplexer, it deduplicates logs within 1 minute, should I change the code to use it?

r0mant replied:
We should rate limit.

tigrato commented on lib/multiplexer/multiplexer.go:
```diff
@@ -414,75 +433,102 @@ func (m *Mux) detect(conn net.Conn) (*Conn, error) {
+					"direct_dst_addr": conn.LocalAddr(),
+					"proxy_src_addr:": newPROXYLine.Source.String(),
+					"proxy_dst_addr:": newPROXYLine.Destination.String(),
+				}).Error(unexpectedPROXYLine)
```
Can you improve the message so it's easier to follow?
Can you please add an actionable message here so users can easily fix their configs?

AntonAM replied:
I tweaked it a little bit, but overall I don't see much problem with the message and it already include actionable part?

tigrato commented on lib/multiplexer/multiplexer.go:
```diff
@@ -414,75 +433,102 @@ func (m *Mux) detect(conn net.Conn) (*Conn, error) {
+					"proxy_src_addr:": newPROXYLine.Source.String(),
+					"proxy_dst_addr:": newPROXYLine.Destination.String(),
+				}).Error(unexpectedPROXYLine)
+				newPROXYLine.Source.Port = 0
```
Add a not why we force the port to be zero for IP  Pinning purposes

tigrato commented on lib/multiplexer/multiplexer.go:
```diff
@@ -414,75 +433,102 @@ func (m *Mux) detect(conn net.Conn) (*Conn, error) {
-				return nil, trace.Wrap(err, invalidProxyLineError)
+
+			if proxyLine != nil {
+				if proxyLine.IsVerified {
```
Should this case ever happen?

capnspacehook replied:
Yeah why shouldn't an error be returned here instead? A signed PROXY header followed by an unsigned PROXY header is allowed?

AntonAM replied:
I think it shouldn't, but I wasn't sure if some combination of load balancers and proxy peering didn't lead to it. And if it would happen that would be correct way to handle it, so didn't return an error to lower probability of unnecessary failures. But thinking again now, I couldn't come up when it would possibly happen so changing it to return an error.

tigrato commented on lib/multiplexer/multiplexer.go:
```diff
@@ -414,75 +433,102 @@ func (m *Mux) detect(conn net.Conn) (*Conn, error) {
+			unsignedPROXYLineReceived = true
+
+			if m.PROXYProtocolMode == PROXYProtocolUnspecified {
+				m.Logger.WithFields(log.Fields{
```
we should rate limit these logs

tigrato commented on lib/authz/permissions.go:
```diff
@@ -365,6 +365,15 @@ func CheckIPPinning(ctx context.Context, identity tlsca.Identity, pinSourceIP bo
+			log.WithFields(logrus.Fields{
+				"client_ip": clientIP,
+				"pinned_ip": identity.PinnedIP,
+			}).Warn("IP pining is not allowed for connections affected by PROXY protocol without explicitly setting 'proxy_protocol: on' in the config.")
```
```suggestion
			}).Warn("IP pining is not allowed for connections behind L4 LoadBalancers with PROXY protocol enabled without explicitly setting 'proxy_protocol: on' in the config.")
```

tigrato commented on lib/authz/permissions.go:
```diff
@@ -365,6 +365,15 @@ func CheckIPPinning(ctx context.Context, identity tlsca.Identity, pinSourceIP bo
 		}
 		return ErrIPPinningMismatch
 	}
+	if clientPort == "0" {
```
Can you please link the places where we replace the port and a short description why we do it

capnspacehook commented on lib/multiplexer/multiplexer.go:
```diff
@@ -414,75 +433,102 @@ func (m *Mux) detect(conn net.Conn) (*Conn, error) {
-			if newProxyLine == nil {
-				continue
+			if newPROXYLine == nil {
+				unsignedPROXYLineReceived = true
```
Given this logic, wouldn't multiple PROXYv2 headers containing LOCAL commands be allowed? Should that be the case?

AntonAM replied:
Yes, but I would say it doesn't matter since LOCAL command doesn't do anything. We can add checking if `unsignedPROXYLineReceived` is already `true` before setting it and returning error if it is, but I'm not sure if it's worth it to complicate code for that.

capnspacehook replied:
I might be worth it to at least log and possibly also emit an event when this happens, as I don't think this should happen in a legitimate scenario? Or could it?

AntonAM replied:
No, it's not a legitimate scenario. I've refactored it to make sure we don't allow multiple PROXY lines with LOCAL command.

capnspacehook commented on lib/multiplexer/multiplexer.go:
```diff
@@ -414,75 +433,102 @@ func (m *Mux) detect(conn net.Conn) (*Conn, error) {
 
 			// This is unsigned proxy line, return error if external PROXY protocol is not enabled
-			if !m.EnableExternalProxyProtocol {
+			if m.PROXYProtocolMode == PROXYProtocolOff {
```
Why isn't this done earlier? Are signed PROXYv2 headers always allowed, even if `proxy_protocol` is set to `off`? That wasn't clear to me from the RFC update.

AntonAM replied:
Yep, signed PROXY headers are always allowed and here, as comment says we enter territory of unsigned header. RFD mentions in the scope that only unsigned PROXY headers are the focus of the RFD.

capnspacehook commented on lib/multiplexer/multiplexer_test.go:
```diff
@@ -256,6 +255,48 @@ func TestMux(t *testing.T) {
 
+	// makes sure the connection gets dropped
+	// when PROXY protocol is 'on' but PROXY line isn't received
+	t.Run("required PROXY line", func(t *testing.T) {
```
Can a test be added for when `proxy_protocol` is also set to `ignored`?

AntonAM replied:
I've added test in the `wrapper_test.go` that checks it, but yeah, I guess I can also add it here more explicitly.

AntonAM committed:
Don't require PROXY headers for connection to itself.

There cases when Telelport will call itself and it can go directly,
avoiding load balancer, so connection will not have unsigned PROXY header.

AntonAM committed:
Refactor validation of 'proxy_protocol' input

AntonAM committed:
Clarify PROXY protocol modes description

AntonAM committed:
Tweak unexpected PROXY protocol line error message

AntonAM committed:
Add comments about setting port to 0

AntonAM committed:
Return an error if we get unsigned PROXY line after signed

AntonAM committed:
Improve wording for error when IP pinning is now allowed when source port = 0

AntonAM committed:
Add test for checking unspecified PROXY protocol mode on multiplexer

capnspacehook commented on lib/multiplexer/multiplexer_test.go:
```diff
@@ -298,6 +298,50 @@ func TestMux(t *testing.T) {
+		// make sure the TLS call failed
+		res, err := utils.RoundtripWithConn(tlsConn)
+		require.Nil(t, err)
+		require.Equal(t, conn.LocalAddr().String(), res)
```
I'm not sure I understand what this test is doing... you upgrade to a TLS connection, and then check that the correct local address was propagated to the backend as the remote address? I'm confused on how that makes sure the TLS call failed

AntonAM replied:
Hehe, I guess it was too late when I worked on it and my mind was somewhere else - I've copied test from the previous one and started changing it for port=0 scenario, but didn't actually finish. 😅 Thanks for catching it.
Fixed in  https://github.com/gravitational/teleport/pull/31622/commits/e5f2e68a6f58c8acaee3afc7eceed7e7b9998bdb

capnspacehook replied:
NP, thanks for fixing!

AntonAM committed:
Fix the test

AntonAM committed:
Refactor tracking of receiving of unsigned PROXY line.

It makes sure we don't allow multiple PROXY lines with LOCAL command.

AntonAM committed:
Improve comment wording.

r0mant commented on lib/multiplexer/multiplexer.go:
```diff
@@ -414,88 +443,125 @@ func (m *Mux) detect(conn net.Conn) (*Conn, error) {
+			proxyLine = newPROXYLine
 			// repeat the cycle to detect the protocol
 		case ProtoTLS, ProtoSSH, ProtoHTTP, ProtoPostgres:
+			selfConnection, err := m.isSelfConnection(conn)
```
We should add a comment here explaining why this check is needed.

r0mant commented on lib/service/kubernetes.go:
```diff
@@ -236,6 +237,7 @@ func (process *TeleportProcess) initKubernetesService(log *logrus.Entry, conn *C
 		DynamicLabels:        dynLabels,
 		CloudLabels:          process.cloudLabels,
 		Log:                  log,
+		PROXYProtocolMode:    multiplexer.PROXYProtocolOff,
```
Can you add a comment here explaining why this is always "off" and doesn't take value from config.

r0mant commented on lib/service/service.go:
```diff
@@ -1929,20 +1929,26 @@ func (process *TeleportProcess) initAuthService() error {
+		log.Info("Starting Auth service with external PROXY protocol support.")
+	}
+	if cfg.Auth.PROXYProtocolMode == multiplexer.PROXYProtocolUnspecified {
+		log.Warn("'proxy_protocol' unspecified." +
```
Update this message to explain how to turn it on or off explicitly in the config.

r0mant commented on lib/srv/alpnproxy/proxy.go:
```diff
@@ -430,12 +430,19 @@ func checkCertIPPinning(tlsConn *tls.Conn) error {
+			p.log.WithFields(logrus.Fields{
+				"client_ip": clientIP,
+				"pinned_ip": identity.PinnedIP,
+			}).Warn("IP pining is not allowed for connections behind L4 load balancers with " +
+				"PROXY protocol enabled without explicitly setting 'proxy_protocol: on' in the config.")
```
```suggestion
			}).Warn("IP pinning is not allowed for connections behind L4 load balancers with " +
				"PROXY protocol enabled without explicitly setting 'proxy_protocol: on' in the proxy_service config.")
```

AntonAM commented on lib/authz/permissions.go:
```diff
@@ -365,6 +365,18 @@ func CheckIPPinning(ctx context.Context, identity tlsca.Identity, pinSourceIP bo
+				"client_ip": clientIP,
+				"pinned_ip": identity.PinnedIP,
+			}).Warn("IP pining is not allowed for connections behind L4 load balancers with " +
+				"PROXY protocol enabled without explicitly setting 'proxy_protocol: on' in the config.")
```
It can be configured in both auth and proxy, that's why I used general `in the config`. Should it be something like `in the Proxy or Auth config`?

r0mant replied:
Yeah, then "... in the proxy_service or auth_service config" or something like that. Basically, we should make the error more actionable so they know what to do exactly.

AntonAM replied:
Done 👍

AntonAM committed:
Fix typo and improve wording.

Co-authored-by: Roman Tkachenko <roman@goteleport.com>

AntonAM commented on lib/config/configuration.go:
```diff
@@ -897,14 +905,29 @@ func applyGoogleCloudKMSConfig(kmsConfig *GoogleCloudKMS, cfg *servicecfg.Config
 	return nil
 }
 
+func validatePROXYProtocolValue(p multiplexer.PROXYProtocolMode) error {
```
I've put it there because it only really affects input file config. Fundamentally `PROXYProtocolModeUnspecified` is not an invalid value for the `PROXYProtocolMode` type.

AntonAM committed:
Clarify PROXYProtocolMode description

AntonAM committed:
Add comment about self connections

AntonAM committed:
Set correct PROXYProtocolMode for multiplexers that don't need unsigned PROXY support

AntonAM committed:
Fix typo.

Co-authored-by: Roman Tkachenko <roman@goteleport.com>

AntonAM committed:
Add comment about PROXY protocol mode on kube service.

AntonAM committed:
Clarify comment for starting auth service in unspecified PROXY protocol mode

r0mant reviewed (approved):
lgtm once remaining comments are addressed

AntonAM committed:
Add dedicated error for rejecting IP pinning on connection with port=0

AntonAM commented on lib/authz/permissions.go:
```diff
@@ -365,6 +365,18 @@ func CheckIPPinning(ctx context.Context, identity tlsca.Identity, pinSourceIP bo
+			}).Warn("IP pining is not allowed for connections behind L4 load balancers with " +
+				"PROXY protocol enabled without explicitly setting 'proxy_protocol: on' in the config.")
+		}
+		return ErrIPPinningMismatch
```
Added `ErrIPPinningNotAllowed` for this case.

AntonAM committed:
Improve error message

AntonAM committed:
Rate limit logging error about unexpected PROXY line.

tigrato commented on lib/authz/permissions.go:
```diff
@@ -336,6 +336,10 @@ var ErrIPPinningMissing = trace.AccessDenied("pinned IP is required for the user
 
+// ErrIPPinningNotAllowed is returned when user's pinned IP doesn't match observed IP.
+var ErrIPPinningNotAllowed = trace.AccessDenied("IP pinning is not allowed for connections behind L4 load balancers with " +
+	"PROXY protocol enabled without explicitly setting 'proxy_protocol: on' in the proxy_service or auth_service config.")
```
Should it be `proxy_service and auth_service config`?

AntonAM replied:
Depends on the client setup, if there's another LB between Proxy and Auth then it can be both, although usually I think it will be only one. I'll make it `and/or`.

AntonAM committed:
Tweak error message.
